code/modules/mob/living/carbon/carbon_update_icons.dm

```diff
@@ -237,27 +237,66 @@
+	var/obj/item/bodypart/l_leg/left_leg
+	var/obj/item/bodypart/r_leg/right_leg
+	var/old_left_leg_key
 	for(var/obj/item/bodypart/limb as anything in bodyparts)
 		limb.update_limb(is_creating = update_limb_data) //Update limb actually doesn't do much, get_limb_icon is the cpu eater.
+
+		if(limb.body_zone == BODY_ZONE_R_LEG)
+			right_leg = limb
+			continue // Legs are handled separately
+
 		var/old_key = icon_render_keys?[limb.body_zone] //Checks the mob's icon render key list for the bodypart
 		icon_render_keys[limb.body_zone] = (limb.is_husked) ? limb.generate_husk_key().Join() : limb.generate_icon_key().Join() //Generates a key for the current bodypart
+
+		if(limb.body_zone == BODY_ZONE_L_LEG)
+			left_leg = limb
+			old_left_leg_key = old_key
+			continue // Legs are handled separately
+
 		if(!(icon_render_keys[limb.body_zone] == old_key)) //If the keys match, that means the limb doesn't need to be redrawn
 			needs_update += limb
 
 
+
+	// Here we handle legs differently, because legs are a mess due to layering code. So we got to process the left leg first. Thanks BYOND.
+	var/list/left_leg_icon // yes it's actually a list, bet you didn't expect that, now did you?
+	var/leg_need_redrawn = FALSE
+	if(left_leg)
+		if(!(icon_render_keys[left_leg.body_zone] == old_left_leg_key))
+			left_leg_icon = left_leg.get_limb_icon()
+			limb_icon_cache[icon_render_keys[left_leg.body_zone]] = left_leg_icon
+			leg_need_redrawn = TRUE
+		else
+			left_leg_icon = limb_icon_cache[icon_render_keys[left_leg.body_zone]]
+
+	if(right_leg)
+		var/old_right_leg_key = icon_render_keys?[right_leg.body_zone]
+		var/left_leg_mask_key = left_leg?.generate_mask_key().Join()
+		right_leg.left_leg_mask_key = left_leg ? left_leg_mask_key : null
+		if(left_leg_mask_key && !right_leg.left_leg_mask_cache[left_leg_mask_key] && left_leg_icon)
+			right_leg.left_leg_mask_cache[left_leg_mask_key] = generate_left_leg_mask(left_leg_icon[1], left_leg_mask_key)
+		icon_render_keys[right_leg.body_zone] = right_leg.is_husked ? right_leg.generate_husk_key().Join() : right_leg.generate_icon_key().Join()
+
+		if(!(icon_render_keys[right_leg.body_zone] == old_right_leg_key))
+			limb_icon_cache[icon_render_keys[right_leg.body_zone]] = right_leg.get_limb_icon()
+			leg_need_redrawn = TRUE
+
+
 	var/list/missing_bodyparts = get_missing_limbs()
 	if(((dna ? dna.species.max_bodypart_count : BODYPARTS_DEFAULT_MAXIMUM) - icon_render_keys.len) != missing_bodyparts.len) //Checks to see if the target gained or lost any limbs.
 		limb_count_update = TRUE
 		for(var/missing_limb in missing_bodyparts)
 			icon_render_keys -= missing_limb //Removes dismembered limbs from the key list
 
-	if(!needs_update.len && !limb_count_update)
+	if(!needs_update.len && !limb_count_update && !leg_need_redrawn)
 		return
 
-	remove_overlay(BODYPARTS_LAYER)
 
 	//GENERATE NEW LIMBS
 	var/list/new_limbs = list()
+	bodyparts = sort_list(bodyparts, /proc/cmp_bodypart_by_body_part_asc)
```

Comment:
Is there a reason this needs to be sorted every time?

Reply:
I couldn't think of any other way to know that the order of the bodyparts changed, and I figured it likely isn't going to take too many operations considering they'll be in order most of the time either way. If you have a better idea on how to do it, I'm all ears.

Reply:
Eh should be fine. God damn fuckin' byond demons making it so i cant make the bodyparts list associative 😭

Reply:
Maybe eventually we can figure a way, it would be *so* much easier, hahaha.